Userland/Libraries/LibWeb/HTML/FormControlInfrastructure.cpp

```diff
@@ -40,4 +46,140 @@ WebIDL::ExceptionOr<Entry> create_entry(JS::Realm& realm, DeprecatedString const
     };
 }
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set
+WebIDL::ExceptionOr<Optional<HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>>>> construct_entry_list(JS::Realm& realm, HTMLFormElement& form)
+{
+    // 1. If form's constructing entry list is true, then return null.
+    if (form.constructing_entry_list())
+        return Optional<HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>>> {};
+
+    // 2. Set form's constructing entry list to true.
+    form.set_constructing_entry_list(true);
+    ScopeGuard scope_guard([&] {
+        form.set_constructing_entry_list(false);
+    });
+
+    // 3. Let controls be a list of all the submittable elements whose form owner is form, in tree order.
+    auto controls = TRY_OR_THROW_OOM(realm.vm(), form.get_submittable_elements());
+
+    // 4. Let entry list be a new empty entry list.
+    HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list = {};
+
+    // 5. For each element field in controls, in tree order:
+    for (auto control : controls) {
+        // 1. If any of the following is true, then continue:
+        // The field element has a datalist element ancestor.
+        if (control->first_ancestor_of_type<HTML::HTMLDataListElement>())
+            continue;
+        // The field element is disabled.
+        if (control->is_actually_disabled())
+            continue;
+        // The field element is a button but it is not submitter.
+        if (is<HTML::HTMLButtonElement>(control.ptr())) {
+            auto* button_element = dynamic_cast<HTML::HTMLButtonElement*>(control.ptr());
+            if (button_element->type() != "submit")
+                continue;
```

Comment:
With submitter they mean the optional argument `submitter` not whether the button is of type submit I think